### PR TITLE
Adds parser task using deep biaffine parser

### LINKED ISSUE #72: Dependency Parsing
# Objective

In its current form, UDTube does not perform dependency parsing. We would love to add it as it is the last part of UD data that we don't use.

This is a hard issue

### PR DESCRIPTION
**Pull request by kylebgorman**
Adds a parser head to the classifier.

This adds a deep biaffine dependency parser head to the classifier. The parser computes head relationships, decoding using the Chu-Liu-Edmonds minimum spanning tree algorithm (greedy decoding produced poor results), classifies each head/dependency arc using the UD dependency relations, and then emits separate cross-entropies for the heads and deprels (these in turn are just summed together with the other tasks losses at present). 

Note that this adds two new hyperparameters: the size of the head/dep and deprel MLPs. We set them to small but sensible default values.

Closely related fixes:

* Parser tests
* Attachment metrics and associated tests
* Improvements to CoNLL-U file parsing
* README updates
* A bibliography file

Drive-by fixes:

* Vocabulary logging is improved.
* A persistent bug in file paths in udtube_test.py is fixed.
* Tests are moved to CPU for greater determinism (even the slowest one runs in <30m).
* Typing is modernized.

Version is bumped to 0.2.0. Closes #72.

### CODE REVIEW

File: udtube/data/conllu.py [hunk 1]
```diff
@@ -153,20 +167,24 @@ def __str__(self) -> str:
     def is_mwe(self) -> bool:
         return self.id_.is_mwe
 
+    @property
+    def is_real(self) -> bool:
+        return not self.id_.is_mwe and not self.id_.is_decimal
```
Comment (DanielYakubov) [inline udtube/data/conllu.py:172]: genuine question, no changes needed, but why not De Morgan here and factor out not?\

`not (self.id_.is_mwe or self.id_.is_decimal)` 

do you find it more readable in this form?

File: udtube/data/mappers.py [hunk 2]
```diff
@@ -48,41 +49,42 @@ def __post_init__(self):
...
+        functor: Callable[[str], int],
     ) -> torch.Tensor:
         """Encodes a tensor.
 
         Args:
-            labels: iterable of labels.
-            vocabulary: a vocabulary.
+            strings: iterable of strings.
+            functor: a callable mapping from strings to integers; usually
+                this is the vocabulary object.
```
Comment (DanielYakubov) [inline udtube/data/mappers.py:60]: Is there ever a case that this won't be a vocab, realistically? I feel like this level of abstraction might be YAGNI, but let me know your thoughts.

My preference is to keep this named vocabulary, but no changes required

File: udtube/data/mappers.py [hunk 3]
```diff
@@ -104,40 +106,70 @@ def encode_lemma(
...
+            Decoded symbols.
         """
         for idx in indices:
-            if idx == special.PAD_IDX:
-                # To avoid sequence length mismatches,
-                # _ is yielded for anything classified as a pad.
-                yield "_"
-            else:
-                yield vocabulary.get_symbol(idx)
+            yield functor(idx)
```
Comment (DanielYakubov) [inline udtube/data/mappers.py:172]: does the functor mapping have an entry for the special PAD_IDX? Is that what allows this clean up?

File: udtube/callbacks.py [hunk 4]
```diff
@@ -110,13 +124,45 @@ def _fill_in_tags(
             except StopIteration:
                 # Prevents the error from being caught by Lightning.
                 logging.error(
-                    f"Length mismatch at tag {attr!r} (sent_id: "
+                    f"Length mismatch at token {token!r} (sent_id: "
+                    f"{tokenlist.metadata.get('sent_id')})"
+                )
+                continue
```
Comment (DanielYakubov) [inline udtube/callbacks.py]: does this continue do anything? There's no logic after it

File: udtube/metrics.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+class LabeledAttachmentScore(AttachmentScore):
+    """Labeled Attachment Score (LAS).
+
+    Computes the percentage of tokens with both correct head assignment
+    and correct dependency relation.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, labeled=True, **kwargs)
```
Comment (DanielYakubov) [inline udtube/metrics.py:100]: Nothing much to say, just point out I thought it was cool how you did this

File: udtube/models.py [hunk 6]
```diff
@@ -196,46 +227,53 @@ def on_train_epoch_end(self) -> None:
...
-            self.feats_accuracy.update(logits.feats, batch.feats)
+        if self.use_parse:
+            head_loss, deprel_loss = self.classifier.parse_head.compute_loss(
+                logits.head,
+                batch.head,
+                logits.deprel,
+                batch.deprel,
+            )
+            # TODO(kbg): maybe something more sophisticated or general is
+            # called for here; test later.
```
Comment (DanielYakubov) [inline udtube/models.py:274]: Are you still thinking about the loss aggregation problem?

File: udtube/parser.py
Comment (DanielYakubov) [inline udtube/parser.py:1]: I reviewed this lightly, mainly just for glaring issues. Lmk if you needed a deeper review, I'd need to dig up some thesis notes to do so.

**Review by DanielYakubov — APPROVED** [review]
Approved with ignorable comments. I'm so glad you came back to do this part!

File: udtube/data/conllu.py (hunk 1 above)
Comment (kylebgorman) [reply udtube/data/conllu.py:172]: A little worse actually...

File: udtube/data/mappers.py (hunk 2 above)
Comment (kylebgorman) [reply udtube/data/mappers.py:60]: It's a specific lambda when encoding the heads, because they're integer (or decimal, or range) IDs.

File: udtube/data/mappers.py (hunk 3 above)
Comment (kylebgorman) [reply udtube/data/mappers.py:172]: Ditto from previous comment.

File: udtube/models.py (hunk 6 above)
Comment (kylebgorman) [reply udtube/models.py:274]: Yeah I may return to this sometime. One possibility is to like do (head_loss + deprel_loss) / 2 so we treat overall parsing loss as having the same magnitude/contribution as the other heads.

File: udtube/callbacks.py (hunk 4 above)
Comment (kylebgorman) [reply udtube/callbacks.py]: Fixed, thanks.
